### PR TITLE
Generalize `geometry.Line`

### PR DESCRIPTION
**Pull request by siefkenj**
This generalizes the functions in `line.py` to work in dimensions higher than 2 where possible.  It is a step along the way to unifying `Line` and `Line3D` in a similar way to the unification of `Point` and `Point3D`.  Incidentally, the generalized routines are shorter, but relying on the `Matrix` object seems to make them slower.  If speed is an issue, optimized routines for particular dimensions can be added after `Line` and `Line3D` are unified.

### CODE REVIEW

**Comment by siefkenj** [comment]
@smichr Could you take a look at this?  I will resolve the new conflicts as soon as I can.

**Comment by smichr** [comment]
Didn't see the ping...I can look at this over the coming week.

**Comment by smichr** [comment]
The way work is going right now I can see this might take longer than this week. If someone else wants to review it, that would be fine with me.

**Comment by siefkenj** [comment]
@smichr do you have time to take a look at this PR yet?

File: sympy/geometry/ellipse.py [hunk 1]
```diff
@@ -835,23 +835,23 @@ def normal_lines(self, p, prec=None):
...
 
         Off-axis points require the solution of a quartic equation. This
         often leads to very large expressions that may be of little practical
         use. An approximate solution of `prec` digits can be obtained by
         passing in the desired value:
 
         >>> e.normal_lines((3, 3), prec=2)
-        [Line(Point2D(-38/47, -85/31), Point2D(9/47, -21/17)),
-        Line(Point2D(19/13, -43/21), Point2D(32/13, -8/3))]
+        [Line2D(Point2D(-0.81, -2.7), Point2D(9/47, -21/17)),
```
Comment (smichr) [inline sympy/geometry/ellipse.py]: Since the evaluated points are Points when the Line is created their decimal approximations are not recomputed to fractions. I think this can be fixed by wrapping the numerically approximated point in `Point`: `points = [Point(pt.n(prec)) for pt in points]` at line 895.

File: sympy/geometry/line.py [hunk 2]
```diff
@@ -36,16 +46,14 @@ class Undecidable(ValueError):
...
-    in a 2-dimensional Euclidean space.
+    """A base class for all linear entities (Line, Ray and Segment)
+    in n-dimensional Euclidean space.
 
     Attributes
     ==========
 
     p1
     p2
-    coefficients
```
Comment (smichr) [inline sympy/geometry/line.py]: Why are these deleted? Beside `p1, p2, and points` there are also `ambient_dimension, coefficients, direction and length`.

File: sympy/geometry/line.py [hunk 3]
```diff
@@ -270,19 +299,25 @@ def is_perpendicular(l1, l2):
...
-        try:
-            a1, b1, c1 = l1.coefficients
-            a2, b2, c2 = l2.coefficients
-            return bool(simplify(a1*a2 + b1*b2) == 0)
-        except AttributeError:
-            return False
+        if not isinstance(l1, LinearEntity) and not isinstance(l2, LinearEntity):
+            raise TypeError('Must pass only LinearEntity objects')
+
+        return l1.direction.dot(l2.direction) == S.Zero
```
Comment (smichr) [inline sympy/geometry/line.py]: Isn't this made weaker by not including a call to `simplify` -- `dot` doesn't do simplification? I wonder if returning an `Eq` which might be unevaluated might be better. One would have to see how this is handled elsewhere.

File: sympy/geometry/line.py [hunk 4]
```diff
@@ -394,19 +443,20 @@ def perpendicular_line(self, p):
...
-        if d2 == 0:  # If a horizontal line
-            if p.y == self.p1.y:  # if p is on this linear entity
-                return Line(p, p + Point(0, 1))
-            else:
-                p2 = Point(p.x, self.p1.y)
-                return Line(p, p2)
-        else:
-            p2 = Point(p.x - d2, p.y + d1)
-            return Line(p, p2)
+        p = Point.pointify(p, dimension=self.ambient_dimension)
```
Comment (smichr) [inline sympy/geometry/line.py]: [that cleaned up nicely]

File: sympy/geometry/line.py [hunk 5]
```diff
@@ -447,26 +497,25 @@ def perpendicular_segment(self, p):
...
-            # ax + by + c = 0
-            y = (-c - a*p.x)/b
-            m = self.slope
-            d2 = 1 + m**2
-            H = p.y - y
-            dx = m*H/d2
-            dy = m*dx
-            p2 = (p.x + dx, y + dy)
+        l = self.perpendicular_line(p)
+        p2, = Intersection(Line(self.p1, self.p2), l)
```
Comment (smichr) [inline sympy/geometry/line.py]: Delete the comma after `p2`?

File: sympy/geometry/line.py [hunk 6]
```diff
@@ -601,43 +604,45 @@ def projection(self, o):
...
+        elif isinstance(other, LinearEntity):
+            p1, p2 = proj_point(other.p1), proj_point(other.p2)
+            # test to see if we're degenerate
+            if p1 == p2:
+                return p1
+            projected = other.__class__(p1, p2)
+            projected = Intersection(self, projected)
+            # if we happen to have intersected in only a point, return that
+            if projected.is_FiniteSet and len(projected) == 1:
+                a, = projected
```
Comment (smichr) [inline sympy/geometry/line.py]: There's that comma again...I don't recognize what its function is.

File: sympy/geometry/line.py [hunk 7]
```diff
@@ -663,141 +668,134 @@ def intersection(self, o):
...
-                    return rv
-                elif isinstance(o, Segment) and inseg(o):
-                    return rv
-            elif isinstance(expr, Segment) and inseg(expr):
-                if isinstance(o, Segment) and inseg(o):
-                    return rv
-            return []
-
-        return o.intersection(self)
+        return other.intersection(self)
```
Comment (smichr) [inline sympy/geometry/line.py]: This is organized well and is clear. I leave the verification to the test suite.

Comment (smichr) [inline sympy/geometry/line.py]: :end: reviewed to here

File: sympy/geometry/line.py (hunk 5 above)
Comment (siefkenj) [inline sympy/geometry/line.py]: `p2, = Intersection(Line(self.p1, self.p2), l)` is a sneaky way of unpacking an iterable with one element.  Since `Intersection` is a set, I cannot do `Intersection(Line(self.p1, self.p2), l)[0]`.  Should I add a comment explaining it is an unpacking, or should I replace it with `p2 = list(Intersection(Line(self.p1, self.p2), l))[0]`?

File: sympy/geometry/line.py (hunk 3 above)
Comment (siefkenj) [inline sympy/geometry/line.py]: Nice idea.  I'll run it through the test suite.

File: sympy/geometry/point.py [hunk 8]
```diff
@@ -172,28 +176,31 @@ def is_collinear(*args):
...
         """
+        args = [Point.pointify(p) for p in args]
 
         # Coincident points are irrelevant; use only unique points.
-        args = list(set(args))
-        if not all(isinstance(p, Point) for p in args):
-            raise TypeError('Must pass only Point objects')
+        # XXX: the current workaround for issue #11238 relies on the points
+        # being in a specific order, so don't let `set` change the order
+        #args = list(set(args))
```
Comment (smichr) [inline sympy/geometry/point.py]: It would be better to leave this as it is and XFAIL the test. Also, does #11554 fix this? Hopefully so, because when this PR is merged with the current master the following fails:

```python
x = 8*tan(13*pi/45)/(tan(13*pi/45) + sqrt(3))
y = (-8*sqrt(3)*tan(13*pi/45)**2 + 24*tan(13*pi/45))/ \
    (-3 + tan(13*pi/45)**2)
assert Line(Point(0, 0), Point(1, -sqrt(3))).contains(Point(x, y)) is True
```

Some orderings of the points works and some don't. I suspect it has to do with the fact that the ratio of `y/x` simplifies to `-sqrt(3)` but the coordinates themselves don't simplify enough for the routine to succeed.

The code you deleted -- using `is_scalar_multiple` works, however. Any reason not to keep it?

File: sympy/geometry/line.py [hunk 9]
```diff
@@ -874,36 +877,19 @@ def random_point(self):
...
-            y = (-c - a*x) / b
-        return Point(x, y)
+        if isinstance(self, Ray):
+            lower = 0
+        if isinstance(self, Segment):
+            lower = 0
+            upper = floor(self.length)
+        t = randint(lower, upper)
+
+        return self.direction*t/abs(self.direction) + self.p1
```
Comment (smichr) [inline sympy/geometry/line.py]: nice!

File: sympy/geometry/point.py [hunk 10]
```diff
@@ -172,28 +176,31 @@ def is_collinear(*args):
...
-        for p in points[1:]:
-            if not Point.is_scalar_multiple(points[0], p):
-                return False
-        return True
+        return Point.affine_rank(*args) == 1
 
-    def is_scalar_multiple(p1, p2):
+    def is_scalar_multiple(self, p2):
         """Returns whether `p1` and `p2` are scalar multiples
         of eachother.
```
Comment (smichr) [inline sympy/geometry/point.py]: Could you find and replace `eachother` -> `each other`?

File: sympy/geometry/line.py [hunk 11]
```diff
@@ -999,33 +1000,19 @@ class Line(LinearEntity):
...
-            else:
-                # go over 1 up slope
-                dx = 1
-                dy = slope
-            # XXX avoiding simplification by adding to coords directly
-            p2 = Point(p1.x + dx, p1.y + dy)
+            ambient_dim = len(p1.p1)
         else:
-            raise ValueError('A 2nd Point or keyword "slope" must be used.')
+            if iterable(p1) and not hasattr(p1, '__len__'):
```
Comment (smichr) [inline sympy/geometry/line.py]: I think you might want to use `is_sequence` instead of `iterable`...probably throughout the PR. Please double check.

File: sympy/geometry/line.py [hunk 12]
```diff
@@ -1057,50 +1044,9 @@ def plot_interval(self, parameter='t'):
...
-        if p1.x == p2.x:
-            return x - p1.x
-        elif p1.y == p2.y:
-            return y - p1.y
-
-        a, b, c = self.coefficients
-        return a*x + b*y + c
-
-    def contains(self, o):
+    def contains(self, other):
```
Comment (smichr) [inline sympy/geometry/line.py]: :end: reviewed to here

File: sympy/geometry/point.py (hunk 8 above)
Comment (smichr) [inline sympy/geometry/point.py]: Never mind -- I see that only 4 of the 6 permutations of point orderings work. But the docstring should be updated to reflect that slopes are not being checked.

File: sympy/geometry/point.py [hunk 13]
```diff
@@ -234,6 +256,95 @@ def ambient_dimension(self):
...
+        For example, if the points lie on a line (and are not all
+        the same) their affine rank is 1.  If the points lie on a plane
+        but not a line, their affine rank is 2."""
+
+        if len(points) == 0:
+            raise TypeError("Must have at least one point")
+
+        # make sure we're genuinely points
+        # and translate to every point to the origin
+        # XXX: issue #11238 we need to pre-simplify otherwise
```
Comment (smichr) [inline sympy/geometry/point.py]: preceding lines:

# make sure all points are Points
# and translate all points to the origin

Regarding the 11238 hacks: could you add a test in the geometry suite which, if it passes, directs someone to go back through geometry and make the appropriate modifications? In case that isn't you, it would be good to include the post-fix line(s) as one or more comments and indicate which line(s) they are replacing.

File: sympy/geometry/point.py [hunk 14]
```diff
@@ -234,6 +256,95 @@ def ambient_dimension(self):
...
+        if isinstance(a, Point):
+            if dimension and a.ambient_dimension != dimension:
+                raise ValueError("Point {} must be of dimension {}.".format(a, dimension))
+            return a
+
+        # any interable that is traversed in a canonical way
+        # (which excludes things like FiniteSets) can be turned
+        # into a point.  This check can be made more general, but for now
+        # exclude known bad objects
+        if iterable(a) and not isinstance(a, (FiniteSet, set, dict)):
```
Comment (smichr) [inline sympy/geometry/point.py]: I suspect `if is_sequence(a):` would work.

File: sympy/geometry/point.py [hunk 15]
```diff
@@ -394,19 +506,22 @@ def intersection(self, o):
...
 
-        return o.intersection(self)
+        return other.intersection(self)
 
     def dot(self, p2):
         """Return dot product of self with another Point."""
-        p2 = Point(p2)
-        return Add(*[a*b for a,b in zip(self, p2)])
+        p2 = Point.pointify(p2, dimension=self.ambient_dimension)
+        # sum could return native python floats, we'd like to ensure the
```
Comment (smichr) [inline sympy/geometry/point.py]: Now that you are using pointify I don't think this "native python floats" situation applies so you can leave the `S()` off.

Comment (smichr) [inline sympy/geometry/point.py]: If there isn't a test of this, one should be added, e.g., one where `p2 = (0.1, 0.2)`

File: sympy/geometry/point.py [hunk 16]
```diff
@@ -442,32 +557,44 @@ def __add__(self, other):
...
-        if iterable(other) and len(other) == len(self):
-            return Point([simplify(a + b) for a, b in zip(self, other)])
-        else:
-            raise ValueError(
-                "Points must have the same number of dimensions")
+        if iterable(other):
+            coords = [simplify(a + b) for a, b in zip(self, other)]
+            if len(coords) != self.ambient_dimension:
+                raise ValueError("Points must have the same number of dimensions")
+            if all(isinstance(c, Number) for c in coords):
```
Comment (smichr) [inline sympy/geometry/point.py]: here and below, I'm not sure this is a good idea...can you explain the rationale of `Point(.1,.2) -> Point(1/10, 1/5)` but `Point(1,2)*0.1 -> Point(0.1, 0.2)`?

File: sympy/geometry/point.py [hunk 17]
```diff
@@ -442,32 +557,44 @@ def __add__(self, other):
...
         coordinates."""
-        return self + (-other)
+        return self + (-x for x in other)
 
     def __mul__(self, factor):
         """Multiply point's coordinates by a factor."""
         factor = sympify(factor)
-        return Point([simplify(x*factor) for x in self.args])
+        coords = [simplify(x*factor) for x in self.args]
+        if all(isinstance(c, Number) for c in coords):
```
Comment (smichr) [inline sympy/geometry/point.py]: see above

File: sympy/geometry/point.py [hunk 18]
```diff
@@ -442,32 +557,44 @@ def __add__(self, other):
...
+            return Point(coords, evaluate=False)
+        return Point(coords)
 
     __truediv__ = __div__
 
     def __neg__(self):
         """Negate the point."""
-        return Point([-x for x in self.args])
+        coords = [-x for x in self.args]
+        if all(isinstance(c, Number) for c in coords):
```
Comment (smichr) [inline sympy/geometry/point.py]: see above

File: sympy/geometry/tests/test_line.py [hunk 19]
```diff
@@ -168,12 +169,11 @@ def test_line_geom():
...
     assert s2.length == sqrt( 2*(x1**2) )
     assert Segment((1, 1), (2, 3)).arbitrary_point() == Point(1 + t, 1 + 2*t)
-    assert s1.perpendicular_bisector() == \
-        Line(Point(1/2, 1/2), Point(3/2, -1/2))
+    assert s1.perpendicular_bisector().equals(Line(Point(1/2, 1/2), Point(3/2, -1/2)))
     # intersections
     assert s1.intersection(Line(p6, p9)) == []
     s3 = Segment(Point(0.25, 0.25), Point(0.5, 0.5))
-    assert s1.intersection(s3) == [s1]
+    assert s1.intersection(s3) == [s3]
```
Comment (smichr) [inline sympy/geometry/tests/test_line.py]: hmm...glad that got fixed

File: sympy/geometry/tests/test_line.py [hunk 20]
```diff
@@ -192,7 +192,7 @@ def test_line_geom():
     assert r4.intersection(r7) == r7.intersection(r4) == [r7]
 
     # Segment contains
-    a, b = symbols('a,b')
+    a, b = symbols('a,b', real=True)
     s = Segment((0, a), (0, b))
     assert Point(0, (a + b)/2) in s
```
Comment (smichr) [inline sympy/geometry/tests/test_line.py]: I don't think we want this to fail when vanilla (non-real) symbols are used. I didn't catch the change that made it so -- is it possible to relax the test so we don't get Undecided?

```python
>>> var('a b')
(a, b)
>>> s = Segment((0, a), (0, b))
>>> Point(0, (a + b)/2) in s
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\geometry\line.py", line 914, in __contains__
    result = self.contains(other)
  File "sympy\geometry\line.py", line 1653, in contains
    raise Undecidable("Cannot determine if {} is in {}".format(other, self))
sympy.geometry.line.Undecidable: Cannot determine if Point2D(0, a/2 + b/2) is in
 Segment2D(Point2D(0, a), Point2D(0, b))
```

File: sympy/geometry/tests/test_line.py [hunk 21]
```diff
@@ -220,10 +220,10 @@ def test_line_geom():
     assert Line((0, 0), (0, 1)).distance(p2) == 1
     assert Line((0, 0), (1, 0)).distance(p1) == 0
     assert Line((0, 0), (1, 0)).distance(p2) == 1
-    m = symbols('m')
+    m = symbols('m', real=True)
     l = Line((0, 5), slope=m)
     p = Point(2, 3)
-    assert l.distance(p) == 2*abs(m + 1)/sqrt(m**2 + 1)
+    assert simplify(l.distance(p) - 2*abs(m + 1)/sqrt(m**2 + 1)).is_zero
```
Comment (smichr) [inline sympy/geometry/tests/test_line.py]: Did the distance routine lose a simplification step? The expression obtained without simplification is `sqrt((3 - (3*m**2 + 2*m + 5)/(m**2 + 1))**2 + (-(-2*m + 2)/(m**2 + 1) + 2)**2)`.

File: sympy/geometry/tests/test_line.py [hunk 22]
```diff
@@ -393,7 +393,7 @@ def test_line3d():
     assert line3.intersection(line4) == [Point3D(2, 2, 1)]
     assert line3.is_parallel(line4) is False
     assert Line3D((0, 1, 2), (0, 2, 3)).intersection(
-        Line3D((0, 1, 2), (0, 1, 1))) == []
+        Line3D((0, 1, 2), (0, 1, 1))) == [Point3D(0,1,2)]
```
Comment (smichr) [inline sympy/geometry/tests/test_line.py]: yes

**Comment by smichr** [comment]
OK, I've made it through. 

- Overall,  I see that this has benefited from mathematical reformulation.

- I didn't like `pointify` at the start, but it grew on me. I think it's the right thing to do.

- I left a few concerns which you have already started to comment on. I will try to check back in this week.

File: sympy/geometry/line.py (hunk 5 above)
Comment (smichr) [inline sympy/geometry/line.py]: It is subtle enough that I think it warrants a comment. It should also probably include a note; `pop` is nice for such cases, too:

```python
p2 = Intersection(Line(self.p1, self.p2), l).pop() # there can only be one point of intersection btwn lines
```

File: sympy/geometry/line.py (hunk 11 above)
Comment (siefkenj) [inline sympy/geometry/line.py]: I did this to be efficient.  Perhaps it is confusing.  `x=[1, 2, 3]` will satisfy `is_sequence(x) == True`, but it will not make it through the if clause since `hasattr(x, '__len__') == True`.  Thus, I avoid an unnecessary call to `tuple(x)`.  Should I be doing a `try` `except` approach instead?

File: sympy/geometry/point.py (hunk 14 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: I avoided `is_sequence` so that `Point(x**2 for x in [1,2,3])` would work.  That is, `g=(x**2 for x in [1,2,3])` passes the `pointify` tests, but `is_sequence(g) == False`.

File: sympy/geometry/point.py (hunk 16 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: I don't know the history of `Point` turning all its arguments into exact types, so I didn't mess with that, but I tried to make it avoid coercing types when it wasn't needed.  (There are some bug's filed about poor numeric performance of `Point`).  I think the best behaviour would be for `Point(.1,.2) -> Point(.1,.2)`.  I suspect this would break a lot of the library though, since things like distance calculations, etc. would then be floats and I think a lot of the library has been programmed with the assumption of exactness...

File: sympy/geometry/tests/test_line.py (hunk 21 above)
Comment (siefkenj) [inline sympy/geometry/tests/test_line.py]: In general, I removed as many calls to `simplify` as I could.  I guess this test should really use `.equals`

File: sympy/geometry/line.py (hunk 2 above)
Comment (siefkenj) [inline sympy/geometry/line.py]: I see.  coefficients should be moved to `Line2D` since it's only relevant there.

File: sympy/geometry/line.py (hunk 5 above)
Comment (siefkenj) [inline sympy/geometry/line.py]: unfortunately, `FiniteSet` doesn't have an attribute `pop`...

File: sympy/geometry/tests/test_line.py (hunk 20 above)
Comment (siefkenj) [inline sympy/geometry/tests/test_line.py]: Good thinking.  I've replaced my containment check with one that relies on the triangle inequality instead of the dot product.

File: sympy/geometry/point.py (hunk 8 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: I think #11554 does fix this.  I tried to update the docstring before and got really frustrated with sphinx errors.  I'll try again.

**Comment by siefkenj** [comment]
I have made several adjustments as per your comments.  When #11554 is merged, I will push them (otherwise a bunch of tests will just fail).

**Comment by smichr** [comment]
#11554 is in.

File: sympy/geometry/point.py (hunk 14 above)
Comment (smichr) [inline sympy/geometry/point.py]: I like the generator instantiation. 

I think `is_sequence(a, include=GeneratorType)` will fit the bill.

File: sympy/geometry/point.py (hunk 16 above)
Comment (smichr) [inline sympy/geometry/point.py]: OK, I'm in agreement. When performing a calculation like `S(1)/3.0` one gets a Float so when multiplying a Point by a float I guess one shouldn't be surprised that it makes the Point inexact.

File: sympy/geometry/point.py (hunk 14 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: I don't know how generous or non-generous we'd like to be, but `is_sequence(iter([1,2,3]), include=GeneratorType) == False`.  We could decide to exclude iterators and only allow sequences or generators, or we could leave the check quite general.  I think passing a long list of exceptions to is_sequence is kindof ugly...(the code for is_sequence isn't terribly complicated.  We'd basically be asking is_sequence to ignore most of its functionality).

Comment (smichr) [inline sympy/geometry/point.py]: Let's insist on literal sequences...you can still use a generator but you have to unpack it yourself which is easy:

```python
>>> def g(*a):
...  print a
...
>>> g(*(i**2 for i in range(3)))
(0, 1, 4)
>>> g(*iter([1,2,3]))
(1, 2, 3)
```

Also, if you could look at the `__new__` routines for `Point` and `pointify` and see if they can be a bit less convoluted (at least that's how it seems to me right now) that would be good. I'll push some ideas to a `generalize` branch if you want to take a look.

Comment (siefkenj) [inline sympy/geometry/point.py]: Sounds good to me.  I'll have a look at this tomorrow.

Comment (siefkenj) [inline sympy/geometry/point.py]: Do you mean `Point2D.__new__` and `Point3D.__new__`?  `Point.__new__` doesn't seem strange to me...though I have no idea why we prevent imaginary coordinates.

File: sympy/geometry/tests/test_point.py [hunk 23]
```diff
@@ -260,3 +260,61 @@ def test_concyclic_doctest_bug():
...
+    test_double = ['is_collinear']
+    for p in singles3d:
+        Point3D(p)
+    for func in test_single:
+        for p in singles3d:
+            getattr(p3d, func)(p)
+    for func in test_double:
+        for p in doubles2d:
+            getattr(p3d, func)(*p)
+    p3d.are_coplanar((1,2,3), [1,2,2], RepeatableGenerator((3,3,2)))
```
Comment (smichr) [inline sympy/geometry/tests/test_point.py]: this line has the wrong syntax and problem should be removed as it is not testing anything

Comment (siefkenj) [inline sympy/geometry/tests/test_point.py]: Can you explain a bit more of what you mean?  That line is supposed to be testing that `are_coplanar` accepts sequence-like inputs.

**Comment by siefkenj** [comment]
What's the status?  Are there things I need to address?

File: sympy/geometry/tests/test_point.py (hunk 23 above)
Comment (smichr) [inline sympy/geometry/tests/test_point.py]: I have made several changes (including this one) which I will send as a friendly PR tomorrow morning.

File: sympy/geometry/point.py [hunk 24]
```diff
@@ -245,18 +214,16 @@ def are_coplanar(*points):
...
-            raise ValueError('must provide 3 or more unique points')
-        return Point.affine_rank(*points) == 2
+        return Point.affine_rank(*points) <= 2
 
     def is_collinear(self, *points):
-        """Is `self` in line with the given sequence of points?
-
-        Test whether or not a set of points is collinear. Returns True if
-        the set of points are collinear and False otherwise.
+        """Retursn `True` if there exists a line
```
Comment (smichr) [inline sympy/geometry/point.py]: Retursn -> Returns

File: sympy/geometry/point.py [hunk 25]
```diff
@@ -245,18 +214,16 @@ def are_coplanar(*points):
...
+        that contains `self` and `points`.  Returns `False` otherwise.
```
Comment (smichr) [inline sympy/geometry/point.py]: You should add the "trivially true" comment here, too.

File: sympy/geometry/point.py [hunk 26]
```diff
@@ -127,93 +149,40 @@ def __new__(cls, *args, **kwargs):
...
-        minimal = not kwargs.get('max', False)
-        dim = kwargs.get('dim',
-            max(i._size for i in points) if minimal else
-            max(i.ambient_dimension for i in points))
-        warn = kwargs.get('warn', False)
-        if not all(a.ambient_dimension == dim for a in points):
-            points = [Point(a, dim=dim, warn=warn) for a in points]
-        return points
+    @classmethod
+    def _normalize_dimension(cls, *args, **kwargs):
```
Comment (smichr) [inline sympy/geometry/point.py]: I preferred to require that the args be Points since the calling routine knows better what is and isn't a Point. The way it is now, if dim is not given then two additional Point calls are being made. It was my impression that Point is rather slow. Perhaps with the mods the speed for handling Point instances is better and this is no longer a significant issue?

Comment (smichr) [inline sympy/geometry/point.py]: I preferred to require that the args be Points since the calling routine knows better what is and isn't a Point. The way it is now, if dim is not given then two additional Point calls are being made. It was my impression that Point is rather slow. Perhaps with the mods the speed for handling Point instances is better and this is no longer a significant issue?

File: sympy/geometry/point.py (hunk 24 above)
Comment (smichr) [inline sympy/geometry/point.py]: Retursn -> Returns

File: sympy/geometry/point.py [hunk 27]
```diff
@@ -404,54 +371,55 @@ def ambient_dimension(self):
...
+        but not a line, their affine rank is 2.  By convention, the empty
+        set has affine rank -1."""
 
         if len(points) == 0:
-            raise TypeError("Must have at least one point")
+            return -1
 
         # make sure we're genuinely points
         # and translate every point to the origin
-        points = Point.normalize_dimensions(*[Point(i) for i in points])
```
Comment (smichr) [inline sympy/geometry/point.py]: If you don't normalize you can get a Matrix error with different length rows: `Point.affine_rank(Point(1,2),Point(1,2,3))`

Comment (smichr) [inline sympy/geometry/point.py]: If you don't normalize you can get a Matrix error with different length rows: `Point.affine_rank(Point(1,2),Point(1,2,3))`

**Comment by smichr** [comment]
Sorry about multiple comments...my browser seems to be having a problem (I have to enter the Preview tab to enter the comment!) and/or I am not understanding how the new review process works.

Note to self: What's the difference between staticmethod and classmethod?

**Comment by smichr** [comment]
Sorry about the conflicts with the new distance calculation. That's just a temporary fix -- go ahead and replace with what we have here.

**Comment by smichr** [comment]
Since all other routines convert non-point args to Point with `evaluate=True` we should think more about whether `Point(1,1)+(.1,.2)` should be `Point(1.1, 1.2)` or `Point(11/10, 6/5)`. The former is consistent with other methods of Point while the latter is more consistent with Number operations.

**Comment by siefkenj** [comment]
I think maybe `Point(1,1)+(.1,.2)` should raise an exception and we should only auto-convert sequences when they are explicitly passed as arguments to methods.

**Comment by smichr** [comment]
> should raise an exception and we should only auto-convert sequences when they are explicitly passed as arguments to methods

OK. `translate` is a method that will do addition. Perhaps we should add a `skew` method that does multiplication: `Point(1,2,3).skew(4) -> Point(4, 2, 3); Point(1,2,3).skew(4, 4, 4) -> Point(4, 8, 12); `.

**Comment by siefkenj** [comment]
We already have `scale` which does the exact same thing.

**Comment by smichr** [comment]
> `scale`

I even looked at that.... OK, nothing new needed. I would leave the current behavior as it is wrt adding non-Points, then. At some point, however, the scale and translate methods should be methods of the general Point rather than only methods of Point2D and Point 3D, perhaps with scale having the signature `(*args, **kwargs)` where kwargs recognizes `pt` as giving the point from which the scaling happens.

**Comment by smichr** [comment]
Are you going to address the issues that were raised in the most recent review?

File: sympy/geometry/point.py (hunk 26 above)
Comment (siefkenj) [inline sympy/geometry/point.py]: Having all coercions to `Point` in one place would ensure that the behaviour is uniform.  The code is likely slightly slower than the old code with `pointify`, but I think it's significantly faster than current master.  Calling `Point(Point(..))` should have almost no overhead. `_normalize_dimension` could be made to avoid the excess calls, but I couldn't think of a way to do it without making the function uglier.

**Comment by siefkenj** [comment]
Most of these issues are fixed in your pull request.  I will merge it as soon as the `on_morph` options make a little more English sense.

**Review by smichr — APPROVED** [review]
I'll commit tomorrow unless there is more that you wanted to do. Thanks for your work on this.

**Comment by siefkenj** [comment]
Great.  Thank you for the help.  Nothing else I was hoping for in this PR. (I was going to alphabetize the methods, but that is best for a different PR).

**Comment by siefkenj** [comment]
Okay...I'm really confused what is wrong with the docstrings to make sphinx upset.  Do you see the error?

**Comment by smichr** [comment]
> Okay...I'm really confused what is wrong with the docstrings to make sphinx upset.  Do you see the error?

No. I twiddled it, too. I don't think we need the `:math:`. You might try to delete the whole thing in an editor and retype it in case there is a gremlin in there. I *have* seen that happen before. Maybe @asmeurer has seen this before.

**Comment by siefkenj** [comment]
I tried deleting the whole docstring and the error has moved to another docstring...

**Comment by smichr** [comment]
> I tried deleting the whole docstring

OK, that's good news. Did it move further down in the file? If so, that means the "start" code that it is recognizing that doesn't have an end code so by doing a binary search on the file contents you should be able to figure out what the start code is. e.g. delete file from halfway above `ambient_dimension` to the end and see if the error shows up. If so, delete the half above that, etc...

**Comment by siefkenj** [comment]
@smichr I am not able to run the sphinx test on my machine.  Are you able to do this bisection?

File: sympy/geometry/point.py [hunk 28]
```diff
@@ -77,82 +107,160 @@ class Point(GeometryEntity):
...
-
-    def is_collinear(*args):
-        """Is a sequence of points collinear?
-
-        Test whether or not a set of points are collinear. Returns True if
-        the set of points are collinear, or False otherwise.
+    @classmethod
+    def are_coplanar(cls, *points):
+        """Return True if there exists a plane in which all the points
+        lie.  A trivial True value is returned if `len(points) < 3 or
```
Comment (jksuom) [inline sympy/geometry/point.py]: len(points) < 3 has only one backquote.

Comment (siefkenj) [inline sympy/geometry/point.py]: Thanks for finding this!

**Comment by smichr** [comment]
OK! It's in.
